backend/src/algorithm/utils/constraint.ts

```diff
@@ -0,0 +1,79 @@
+import { maxSubjectLecturesPerDay } from '../consts';
+
+function checkIfClassUnavailable(timetable: MatrixHashmap, class_: Class, dayIndex: number, periodIndex: number) {
+  return timetable[hash(class_.name)][dayIndex][periodIndex];
+}
+
+function checkIfTeacherUnavailable(unavailable: Unavailable, subject: Subject, dayIndex: number, periodIndex: number) {
+  if (subject === null) return;
+  return unavailable.teachers[hash(subject.teacher)][dayIndex][periodIndex];
+}
+
+function checkIfClassroomUnavailable(
+  unavailable: Unavailable,
+  subject: Subject,
+  dayIndex: number,
+  periodIndex: number,
+) {
+  if (subject === null) return;
+  return subject.classroom && unavailable.classrooms[hash(subject.classroom)][dayIndex][periodIndex];
+}
+
+function checkIfLectureQuantityFulfilled(remainingLectures: RemainingLectures, class_: Class, subject: Subject) {
+  if (subject === null) return;
+  return remainingLectures[class_.name][subject.name] === 0;
+}
+
+function checkIfDailyLimitExceeded(timetable: MatrixHashmap, class_: Class, subject: Subject, dayIndex: number) {
+  if (subject === null) return;
+  const lectureCount = getLectureCountForDay(timetable, class_, subject, dayIndex) as number;
+  return lectureCount >= maxSubjectLecturesPerDay;
+}
+
+export function checkConstraints(
+  timetable: MatrixHashmap,
+  unavailable: Unavailable,
+  remainingLectures: RemainingLectures,
+  class_: Class,
+  subject: Subject,
+  dayIndex: number,
+  periodIndex: number,
+) {
+  return (
+    checkIfClassUnavailable(timetable, class_, dayIndex, periodIndex) || // class in not available
+    checkIfTeacherUnavailable(unavailable, subject, dayIndex, periodIndex) || // teacher is not available
+    checkIfClassroomUnavailable(unavailable, subject, dayIndex, periodIndex) || // classroom is not available
+    checkIfLectureQuantityFulfilled(remainingLectures, class_, subject) || // all lectures for subject are in schedule
+    checkIfDailyLimitExceeded(timetable, class_, subject, dayIndex) // max lecture daily quantity is exceeded
+  );
+}
+
+export function validateSwap(
+  timetable: MatrixHashmap,
+  unavailable: Unavailable,
+  remainingLectures: RemainingLectures,
+  class_: Class,
+  newSubject: Subject,
+  newDayIndex: number,
+  newPeriodIndex: number,
+  oldSubject: Subject,
+  oldDayIndex: number,
+  oldPeriodIndex: number,
+) {
+  return (
+    // validate if old subject can be set to new time slot
+    !(
+      checkIfClassUnavailable(timetable, class_, newDayIndex, newPeriodIndex) ||
+      checkIfTeacherUnavailable(unavailable, oldSubject, newDayIndex, newPeriodIndex) ||
+      checkIfClassroomUnavailable(unavailable, oldSubject, newDayIndex, newPeriodIndex)
+    ) &&
+    // validate if new subject can be set to previously set time slot
+    !(
+      checkIfTeacherUnavailable(unavailable, newSubject, oldDayIndex, oldPeriodIndex) ||
+      checkIfLectureQuantityFulfilled(remainingLectures, class_, newSubject)
+    )
+  );
```

Comment:
This is even more difficult to read and parse mentally. One of the reasons is the negation of a couple conditions that are wrapped in brackets. The de Morgans rule should be applied there.
Generally try to "dumb" down the code as much as possible so that you or somebody else can read it more easily in a week, month, year, etc. In my opinion, something like this is easier to read and understand:
```js
if (checkIfClassUnavailable(timetable, class_, newDayIndex, newPeriodIndex)) return false;
if (checkIfTeacherUnavailable(unavailable, oldSubject, newDayIndex, newPeriodIndex)) return false;      
if (checkIfClassroomUnavailable(unavailable, oldSubject, newDayIndex, newPeriodIndex)) return false;
if (checkIfTeacherUnavailable(unavailable, newSubject, oldDayIndex, oldPeriodIndex)) return false;
if (checkIfLectureQuantityFulfilled(remainingLectures, class_, newSubject)) return false;

return true;     
```

Maybe the return statements need to be flipped (the ifs return true while the main one is false), but I believe the basic implication is correct; that code formatted like this is much easier to follow and understand.

Reply:
As I mentioned before, this approach would never cross my mind. Great advice, thank you once again.